lstchain/high_level/interpolate.py

```diff
@@ -0,0 +1,577 @@
+from scipy.interpolate import griddata
+
+log = logging.getLogger(__name__)
+
+
+def interp_params(params_list, data):
+    """
+    From a given list of angular parameters, to be used for interpolation,
+    take values from a given data table/dict.
+
+    Returns the neccessary values with applied functions as need be for the
+    interpolation, for each parameter as a list.
+    """
+    mc_pars = []
+    if "ZEN_PNT" in params_list:
+        mc_pars.append(np.cos(u.Quantity(data["ZEN_PNT"], "deg").to_value(u.rad)))
+
+    if "B_DELTA" in params_list:
+        mc_pars.append(np.sin(u.Quantity(data["B_DELTA"], "deg").to_value(u.rad)))
+
+    if "AZ_PNT" in params_list:
+        mc_pars.append(u.Quantity(data["AZ_PNT"], "deg").to_value(u.rad))
+
+    return mc_pars
+
+
+def check_in_delaunay_triangle(irfs, data_params, use_nearest_irf_node=False):
+    """
+    From a given list of IRFs as grid points used for interpolation, retrieve
+    the Delaunay triangulation list of IRFs, where the simplex includes the
+    target points in data_params.
+
+    If the target point does not exist inside the simplex, the IRF
+    corresponding to the nearest grid point to the target value. This is also
+    acheived if use_nearest_irf_node is passed as True.
+
+    Fetching the nearest IRF node is also performed by checking the azimuth
+    values, as there might be more than a single node, with the same position
+    in the interpolation parameter space.
+
+    If the list of given IRFs are not enough for calculating the Delaunay
+    triangulation, an empty list is returned.
+
+    Parameters
+    ----------
+    irfs: List of IRFs to check for Delaunay triangulation.
+        'List'
+    data_pars: Dict of arrays of range of parameters of the observed data
+        in the event list, to check for interpolation.
+        'Dict'
+    use_nearest_irf_node: Boolean if we need only the nearest IRF node.
+        'Bool'
+
+    Returns
+    -------
+    irf_list: Revised list of IRFs after the checks.
+        'List'
+    """
+    # Exclude AZ_PNT as target interpolation parameter
+    d = data_params.copy()
+    d.pop("AZ_PNT", None)
+    # For the interpolation parameters only
+    data_pars_sel = [*d.keys()]
+
+    new_irfs = []
+    mc_params_sel = np.empty((len(irfs), len(data_pars_sel)))
+    mc_params_full = np.empty((len(irfs), len(data_params)))
+
+    for i, irf in enumerate(irfs):
+        f = fits.open(irf)[1].header
+
+        mc_pars_sel = interp_params(data_pars_sel, f)
+        mc_params_sel[i, :] = np.array(mc_pars_sel)
+
+        mc_pars_full = interp_params(data_params, f)
+        mc_params_full[i, :] = np.array(mc_pars_full)
+
+    data_val_sel = interp_params(data_pars_sel, data_params)
+    data_val_full = interp_params(data_params, data_params)
+
+    try:
+        tri = Delaunay(mc_params_sel)
+    except ValueError:
+        print("Not enough grid values for Delaunay triangulation")
+        return new_irfs
```

Comment:
Does this mean that your code prevents any DL3 production when not providing at least 3 IRF files?
If so I would say you need to change this. At least if use_nearest_irf_node is True it should proceed with this method (trivial if one IRF, same as your current implementation if 2).

Also looking in lstchain/tools/lstchain_create_dl3_file.py line 331 it seems like it fills you list with [] if you use a single IRF file. While the following code assume it contains a single path.

Reply:
In the DL3 Tool, https://github.com/cta-observatory/cta-lstchain/blob/78223b7541e7c20d55c145242927487eb17ed823/lstchain/tools/lstchain_create_dl3_file.py#L287-L311 I already take into account of the user passing only a single IRF, or if the glob search in the IRF directory path results in only a single IRF. In that case, no interpolation process is initiated.

The exception here is based on the minimum requirement on the number of nodes (n+1) to create a convex hull based on the interpolation parameters (n). In the stereo IRF interpolation, this will require at least 4 nodes to create an interpolation grids in 3-parameters space.

Reply:
Ok so only my issue with the number of nodes being between 2 and n remains. If you use nearest node, even if Delaunay fails you should be able to create DL3

Reply:
Ok, I understand your point. I will fix to cover that option.

Reply:
Good. Honestly it will probably never happen. But I think you can handle it without making the code more complex